File: Verify/Java/Messaging/src/com/dynmark/Verify.java
package com.dynmark;

import java.io.IOException;
import java.util.Scanner;
import java.util.UUID;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.mashape.unirest.http.HttpResponse;
import com.mashape.unirest.http.Unirest;
import com.mashape.unirest.http.exceptions.UnirestException;

public class Verify {
	private static String userName = "@username@";
	private static String password = "@Password@";
	private static UUID requestId;
	private static ObjectMapper objectMapper = new ObjectMapper();

	private static boolean SendCode(String phoneNumber) throws UnirestException, IOException {
		HttpResponse<String> response = 
				Unirest.post("https://services.dynmark.com/webapi/verify")
				.header("content-type", "application/json")
				.header("accept", "application/json")
				.basicAuth(userName, password)
				.body("{"
						+ "'from': 'Example',"
						+ "'to': '" + phoneNumber + "',"
						+ "}")
				.asString();

		if (response.getStatus() == 201)
		{
			SendCodeResponse body = objectMapper.readValue(response.getBody(), SendCodeResponse.class);
			requestId = body.getRequestId();
			return true;
		}
		else if (response.getStatus() == 401)
		{
			System.out.println("Your username and password are incorrect");
			return false;
		}
		else if (response.getStatus() == 400)
		{
			ApiValidationFailure[] failures = objectMapper.readValue(response.getRawBody(), ApiValidationFailure[].class);

			System.out.println("Sorry, we couldn't send you a code.");
			for (ApiValidationFailure failure : failures)
			{
				if (failure.getFailureCode().equals("ToInvalid"))
				{
					System.out.println("Your phone number doesn't look like a valid number.");
				}
			}

			return false;
		}
		else
		{
			System.out.println("Non success response " + response.getStatus());
			return false;
		}
	}

	private static boolean ValidateCode(String code) throws UnirestException, IOException {
		HttpResponse<String> response = 
				Unirest.put("https://services.dynmark.com/webapi/verify/" + requestId.toString() + "/validate")
				.header("content-type", "application/json")
				.header("accept", "application/json")
				.basicAuth(userName, password)
				.body("{"
						+ "'code': '" + code + "'"
						+ "}")
				.asString();

		if (response.getStatus() == 200)
		{
			ValidateResponse body = objectMapper.readValue(response.getBody(), ValidateResponse.class);

			if (body.getStatus().equals("CodeVerified"))
			{
				System.out.println("Your code was correct and is verified.");
				return true;
			}

			System.out.println("Sorry, that code was not recognised. Reason: " + body.getStatus());
			return false;
		}
		if (response.getStatus() == 204)
		{
			System.out.println("Request id not recognised");
			return true;
		}
		else if (response.getStatus() == 401)
		{
			System.out.println("Your username and password are incorrect");
			return false;
		}
		else if (response.getStatus() == 400)
		{
			ApiValidationFailure[] failures = objectMapper.readValue(response.getRawBody(), ApiValidationFailure[].class);

			System.out.println("Sorry, we couldn't send you a code.");
			for (ApiValidationFailure failure : failures)
			{
				if (failure.getFailureCode().equals("ToInvalid"))
				{
					System.out.println("Your phone number doesn't look like a valid number.");
				}
			}

			return false;
		}
		else
		{
			System.out.println("Non success response " + response.getStatus());
			return false;
		}
	}

	public static void main(String[] args) throws UnirestException, IOException {
		Scanner scanIn = new Scanner(System.in);
		System.out.println("Enter your phone number: ");
		if (SendCode(scanIn.nextLine()))
		{
			System.out.println("Enter the code you received: ");
			while (!ValidateCode(scanIn.nextLine()))
			{
			}
		}

		scanIn.close();         
	}
}

File: Verify/Node/routes/verify.js
var express = require('express');
var request = require('request-promise');
var router = express.Router();

global.verifyRequestId = null;

const _useranme = "**Your username**";
const _password = "**Your password**";

router.get('/', function(req, res){
    res.redirect('/register');
  });

router.get('/register', function(req, res, next) {
    res.render('register');
});

router.post('/register', function(req, res, next) {
    var phoneNumber = req.body.phoneNumber;
    if (phoneNumber)
    {
        request({
            method: 'POST',
            url: 'https://services.dynmark.com/webapi/verify',
            auth: {
                'user': _useranme,
                'pass': _password
            },
            timeout: 130000,
            json: true,
            body: {
                from: "Example",
                to: phoneNumber
            }
        })
        .then(function(body) {
            if (body.numberStatus === "On" || body.numberStatus === "Off")
            {
                global.verifyRequestId = body.requestId;
                res.redirect("verify");
                res.end();
            }
            else
            {
                res.render('register', {message:"Sorry, that number appears to be dead."});
            }
        })
        .catch(function(err)
        {
            var errors = "";
            if (err.statusCode == 401)
            {
                errors = "Invalid credentials";
            }
            else if (err.statusCode == 400)
            {
                if (err.error.some(r => r.failureCode == "ToInvalid"))
                {
                    errors += "Your phone number doesn't look like a valid number.";
                }
                if (err.error.some(r => r.failureCode == "Throttled"))
                {
                    errors += "Please wait for 30 seconds before requesting a resend.";
                }
            }
            else
            {
                errors = "Non success response";
            }
            
            res.render('register', {message:errors});
        })
        .finally(function()
        {
        });
    }
    else
    {
        res.render('register');
    }
});


router.get('/verify', function(req, res, next) {
    res.render('verify');
});

router.post('/verify', function(req, res, next) {
    var code = req.body.code;
    if (code && global.verifyRequestId)
    {
        request({
            method: 'PUT',
            url: 'https://services.dynmark.com/webapi/verify/' + global.verifyRequestId + "/validate",
            auth: {
                'user': _useranme,
                'pass': _password
            },
            timeout: 130000,
            json: true,
            body: {
                code: code
            }
        })
        .then(function(body) {
            if (body && body.status == "CodeVerified")
            {
                res.render('verify', {message:"Your code was correct and is verified."});
            }
            else
            {
                res.render('verify', {message:"Sorry, that code was not recognised. Reason " + (body ? body.status : "CodeUnrecognised")});
            }
        })
        .catch(function(err)
        {
            var errors = "";
            if (err.statusCode == 401)
            {
                errors = "Invalid credentials";
            }
            else if (err.statusCode == 400)
            {
                errors = "Please correct errors";
            }
            else
            {
                errors = "Non success response";
            }
            
            res.render('verify', {message:errors});
        })
        .finally(function()
        {
        });
    }
    else
    {
        res.render('verify');
    }
});

module.exports = router;
